Address comments from Ilari (#28)
ietf-wg-jose/draft-ietf-jose-hpke-encrypt: draft-ietf-jose-hpke-encrypt.md

# Key Encryption

HPKE based recipients can be added alongside existing `ECDH-ES+A128KW` or `RSA-OAEP-384` recipients because HPKE is only used to encrypt the content encryption key, and because the protected header used in content encryption is passed to HPKE as Additional Authenticated Data.

> This protected header is formatted as BASE64URL(UTF8(JWE Protected Header)).

Reviewer: ```suggestion
```

This sentence is redundant.  Please delete it.  This is already defined by JWE.

Author: It was to address the comment from Illari on "How is HPKE AAD input set"
I can re-phrase the line to say 
The protected header encoding remains consistent with existing JWE formatting rules.

In HPKE JWE Key Encryption:

> - The JWE protected header MUST NOT contain an "alg". In HPKE, the algorithm used to encrypt the CEK is specified per-recipient in the recipient's header, making it unnecessary to include it in the JWE protected header.

Reviewer 2: ```suggestion
- The JWE protected header MUST NOT contain the "alg", when recipients use different algorithms to secure the content encryption key.

- The JWE protected header SHOULD contain the "alg", when all recipients use the same HPKE algorithm to secure the content encryption key.
```

Reviewer: https://www.rfc-editor.org/rfc/rfc7516.html#section-7.2.1 says:
> Some Header Parameters, including the "alg" parameter, can be shared among all recipient computations.  Header Parameters in the JWE Protected Header and JWE Shared Unprotected Header values are shared among all recipients.

Orie is right that we cannot prohibit "alg" in the protected header.

> - The JWE Protected Header MUST include an "enc" value registered in both the IANA HPKE AEAD Identifiers Registry and the IANA JSON Web Signature and Encryption Algorithms Registry to ensure consistency across recipients (see Section 7.2.1 of {{RFC7516}}).

Reviewer 2: https://datatracker.ietf.org/doc/html/rfc7516#section-7.2.1

Not sure about the "see" reference here.

Is it needed.

Author: I am referring to the following text Section 7.2.1 of RFC 7516:

   This primarily means that the "enc" (encryption algorithm) Header
   Parameter value in the JOSE Header for each recipient and any
   parameters of that algorithm MUST be the same.

> - The JWE Protected Header MUST include an "enc" value registered in both the IANA HPKE AEAD Identifiers Registry and the IANA JSON Web Signature and Encryption Algorithms Registry to ensure consistency across recipients (see Section 7.2.1 of {{RFC7516}}).

Reviewer 2: ```suggestion
- The JWE Protected Header MUST include an "enc" value registered in both the IANA HPKE AEAD Identifiers Registry and the IANA JSON Web Signature and Encryption Algorithms Registry.
```

Reviewer 2: Its obvious that both registries are needed, you can't understand HPKE without the first, and you can't understand JWE without the second.

- The recipient unprotected header parameters "psk_id" and "auth_kid" MAY be present.
- The recipient unprotected header parameter "ek" MUST be present.
- The recipient unprotected header MUST contain a registered HPKE "alg" value.
- Recipient JWE Encrypted Key MUST be the ciphertext from HPKE Encryption.
- The "encrypted_key" MUST be the base64url encoded JWE Encrypted Key as described in Step 15 in Section 5.1 of {{RFC7516}}. The recipient 
  "encrypted_key" is as described in Section 7.2.1 of {{RFC7516}}.
- If the selected AEAD algorithm requires an IV and authentication tag, the "iv" and "tag" parameters MUST be present, as specified in 
  {{RFC7516}}.
- The "aad" JWE member MAY be present.
- The "ciphertext" MUST be the base64url encoded ciphertext as described in Step 16 in Section 5.1 of {{RFC7516}}.

> - The HPKE Setup info parameter MUST be set to an empty string.

Reviewer 2: Not sure this needs to be a MUST... it SHOULD be empty string.

Author: The COSE-HPKE draft specifies that the setup info is set to an empty string. We should ensure alignment between both drafts. In a meeting, Mike suggested that if auxiliary authenticated application information needs to be provided to HPKE setup info, a new JOSE parameter should be introduced.

Reviewer 3: This should be addressed in a separate PR.

> - The HPKE Setup info parameter MUST be set to an empty string.

Reviewer 2: ```suggestion
- The HPKE Setup info parameter SHOULD be set to an empty string.
```

Author: The COSE-HPKE draft specifies that the setup info is set to an empty string. We should ensure alignment between both drafts. In a meeting, Mike suggested that if auxiliary authenticated application information needs to be provided to HPKE setup info, a new JOSE parameter should be introduced.

Reviewer 2: https://github.com/cose-wg/HPKE/issues/68

Reviewer: I think the ultimate language we want uses MUST, saying something like:
> The HPKE Setup info parameter MUST be set to the empty string unless set with the YET-TO-BE-DEFINED header parameter.

Author: Yes, a new YET-TO-BE-DEFINED header parameter sounds good for HPKE Setup info. We can take it up as a separate PR.

Reviewer 2: I don't agree. If HPKE Setup info is set it can be any of the following:

1. Context, like in concatKDF.
2. private information known to both parties.
2. public information included in the message (like a header parameter)

I don't think we need to specify any of this, except perhaps to state that "apu" and "apv" are not required to be processed when present.
